### ISSUE TITLE
Question about the UV mask ground truth

### ISSUE THREAD

**Issue by suyz526** [yshah43/DPOD#4]
Hi,

thanks a lot for this repository.

I have a question about the UV mask ground truth. The output of U channel and V channel have the same size as the input image. So it will also have a prediction for each background pixel.  Then how to define the ground-truth value of the background pixel in the UV mask ground-truth? From the code, looks like it is defined as 0?

And the loss function for the U and V, it also punishes the prediction in the background pixels. However, background pixels don't have the correspondence with the 3D Model. Should the loss function for the U and V punish only with prediction within the predicted mask?

Thanks in advance.

**Comment by yshah43** [comment]
Yes the background pixels for UV masks are defined as 0. 

The background pixels are being punished because then only the correspondence block will be able to learn where the object is. But the way you're thinking is interesting, one major obstacle is when the input is just an RGB image how will you know what part of the UV mask has the object? (Remember ID and UV masks are all being trained together at the same time).

**Comment by suyz526** [comment]
> Yes the background pixels for UV masks are defined as 0.
> 
> The background pixels are being punished because then only the correspondence block will be able to learn where the object is. But the way you're thinking is interesting, one major obstacle is when the input is just an RGB image how will you know what part of the UV mask has the object? (Remember ID and UV masks are all being trained together at the same time).

Thanks a lot for your reply.

So, in the UV mask ground-truth, the background is 0. And at the same time, (0,0) also related to a point in the 3D model. This sounds a bit ambiguous to me. But I think what you do is the same way as the author did.

For the loss function, I'm thinking maybe we can make use of the id-mask output in the UV loss. If we can turn id-mask as a binary mask, then we can elementwise product this binary mask to the UV mask. After that, we calculate the loss of the UV mask.

In the case of single object prediction, the id-mask is a binary mask, which could be work.

**Comment by yshah43** [comment]
That actually makes sense. If the UV masks have better output that would improve the initial pose estimates via PnP which was causing a lot of issues.

**Comment by suyz526** [comment]
Thanks for the friendly reply. 

There are 2 more details I want to discuss with you.

1. In `train_correspondence_block`, `correspondence_block = UNET.UNet(
        n_channels=3, out_channels_id=16, out_channels_uv=256, bilinear=True)` 
could the `out_channels_id` be 14? (13 different object + background) 

2. From the paper, the decoder for U channel and V channel are separate (Fig2). In your code, the decoder for U channel and V channel shares weights?

**Comment by yshah43** [comment]
> There are 2 more details I want to discuss with you.
> 
> 1. In train_correspondence_block, correspondence_block = UNET.UNet( n_channels=3, out_channels_id=16, out_channels_uv=256, bilinear=True)
> could the out_channels_id be 14? (13 different object + background)
> 
> 2. From the paper, the decoder for U channel and V channel are separate (Fig2). In your code, the decoder for U channel and V channel shares weights?

1. LineMOD dataset has 15 objects and I realized at the end that DPOD only evaluated for 13 of those 15 objects which is why I might have forgotten to change it. out_channels_id should be 14. Thanks for catching that error!

2. No, they don't share weights. Check the forward function in Unet_model.py.

**Comment by suyz526** [comment]
> > There are 2 more details I want to discuss with you.
> > 
> > 1. In train_correspondence_block, correspondence_block = UNET.UNet( n_channels=3, out_channels_id=16, out_channels_uv=256, bilinear=True)
> >    could the out_channels_id be 14? (13 different object + background)
> > 2. From the paper, the decoder for U channel and V channel are separate (Fig2). In your code, the decoder for U channel and V channel shares weights?
> 
> 1. LineMOD dataset has 15 objects and I realized at the end that DPOD only evaluated for 13 of those 15 objects which is why I might have forgotten to change it. out_channels_id should be 14. Thanks for catching that error!
> 2. No, they don't share weights. Check the forward function in Unet_model.py.
> 
> EDIT: 1 has been corrected.

I'm a mxnet user, and I have some trouble to run the code on my machine, but I'm working on it.

At least in mxnet, it should be something like:

#U Mask
        self.up1_u = Up(1024, 512, bilinear)
        self.up2_u = Up(512,512,bilinear)
        self.outc_u1 = OutConv(256, out_channels_uv)
        self.outc_u2 = OutConv(256, out_channels_uv)
        self.outc_u3 = OutConv(256, out_channels_uv)
        self.outc_u4 = OutConv(256, out_channels_uv)
        self.up3_u = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
        self.up4_u = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)

#V Mask
        self.up1_v = Up(1024, 512, bilinear)
        self.up2_v = Up(512,512,bilinear)
        self.outc_v1 = OutConv(256, out_channels_uv)
        self.outc_v2 = OutConv(256, out_channels_uv)
        self.outc_v3 = OutConv(256, out_channels_uv)
        self.outc_v4 = OutConv(256, out_channels_uv)
        self.up3_v = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
        self.up4_v = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)



Maybe in Pytorch is different, but may you have a check, if the logits_u and logits_v always have the same output? (logits_u - logits_v =0)

**Comment by yshah43** [comment]
Actually you're right. I'll fix it
